PyInstaller/hooks/hook-numpy.py

```diff
@@ -1,54 +1,159 @@
+    # NumPy remains compatible with PyInstaller 4.0, I'm copying it here.
+
+    names = set()
+    binaries = []
+    for folder in os.environ["PATH"].split(os.pathsep):
+        for path in Path(folder).glob(name):
+            if not path.name in names:
+                binaries.append((str(path), "."))
+                names.add(path.name)
+    if not binaries:
+        logger.warning(
+            "Failed to find '%s' DLL in PATH. Your app will likely crash if run"
+            " on a different machine that doesn't already have it.", name
         )
-    extend_system_path(numpy_core_paths)
-    del numpy_core_paths
-
-# if we bundle the testing module, this will cause
-# `scipy` to be pulled in unintentionally but numpy imports
-# numpy.testing at the top level for historical reasons.
-# excludedimports = collect_submodules('numpy.testing')
-
-binaries = []
-
-# package the DLL bundle that official numpy wheels for Windows ship
-# The DLL bundle will either be in extra-dll on windows proper
-# and in .libs if installed on a virtualenv created from MinGW (Git-Bash
-# for example)
-if is_win:
-    extra_dll_locations = ['extra-dll', '.libs']
-    for location in extra_dll_locations:
-        dll_glob = os.path.join(os.path.dirname(
-            get_module_file_attribute('numpy')), location, "*.dll")
-        if glob.glob(dll_glob):
-            binaries.append((dll_glob, "."))
+    return binaries
+
+
+if is_mkl:
+    # Other dlls that MKL uses which PyInstaller can't detect itself. These are
+    # determined empirically using dynamic dependency sniffing and will lead
+    # to issues in the future should they change...
+    [binaries.extend(find_library(i))
+     for i in ["libcrypto*", "libffi*", "libssl*"]]
+
+
+# --- A vain attempt at Conda's numpy support ---
+
+# Regular numpy, even with unofficial mkl builds, is pretty trivial to support
+# with PyInstaller. Unfortunately Conda's numpy is the opposite. We need
+# their help to maintain this because our own attempts have been a disaster.
+
+if compat.is_conda:
+    hiddenimports.append("six")
+
+    # There are so many hidden binary dependencies. This list is heavily, OS,
+    # Python and NumPy versions dependent. Omitting any of these can lead to
+    # obscure and often traceback-less crashes.
+    # XXX: As you can see, this is really not a scalable solution. Needs help!
+    conda_dll_patterns = [re.compile(i) for i in (
+        'apphelp.*', 'crypt32.*', 'imagehlp.*', 'libblas.*',
+        'libcblas.*', 'libcrypto.*', 'libffi.*', 'libgcc_.*',
+        'libgfortran.*', 'libifcoremd.*', r'libiomp\d+md.*', 'liblapack.*',
+        'libmmd.*', 'libomp.*',
+        'libopenblas.*', 'libquadmath.*', 'libssl.*', 'libuuid.*',
+        'libz.*', 'mkl_avx.*', 'mkl_core.*',
+        'mkl_intel_thread.*', 'mkl_rt.*', 'mkl_vml_avx.*',
+        'mkl_vml_avx.*', 'msasn.*', 'mswsock.*', 'ole.*',
+        'oleaut.*', 'tbbmalloc.*', 'urandom'
+    )]
```

Comment:
These wildcards are causing debug (`.pdb`) to be probed on Windows+anaconda, which raises an error dialog, with message like:
```
C:\Users\rok\anaconda3\envs\pyinstaller-numpy-conda\Library\bin\libcrypto-1_1-x64.pbd
is either not designed to run on Windows or it contains an error. Try installing the
program again using the original installation media or contact your system administrator
or the software vendor for support. Error status 0xc000012f.
```
After clicking OK, the build resumes normally, but this happens for `libcrypto-1_1-x64.pdb`, `libssl-1_1-x64.pdb`, and `libiomp5md.pdb` (actually twice for `libcrypto-1_1-x64.pdb` and `libssl-1_1-x64.pdb`, because they are probed both here and under `is_mkl`).

I supposed `.pdb` suffix should be explicitly excluded, or valid dynamic libraries suffices should be specified instead.

Reply:
Eww gross. So some of those dlls have unusual suffixes which is why I didn't just use an explicit list of them. Unsurprisingly this drags all kinds of junk in. I was hoping my [lazy *is this a dll* check](https://github.com/pyinstaller/pyinstaller/blob/998e4057d0c0a99dee35bbabe26602df84a434a3/PyInstaller/hooks/hook-numpy.py#L151) would clear this up for me but evidently it's having side effects. Unreassuringly I don't see this issue when I test it either locally or in CI despite the `pbd`s being present (add to list of reasons why Conda is untestable). Can you share your environment details?

<details><summary>The original list I'm working from is here</summary>

```
urandom
libcrypto.1.0.0.dylib
libssl.1.0.0.dylib
libgcc_s.1.dylib
libgfortran.3.dylib
libopenblasp-r0.3.5.dev.dylib
libquadmath.0.dylib
libcrypto.1.1.dylib
libssl.1.1.dylib
libffi.6.dylib
libopenblasp-r0.3.3.dylib
libgfortran.4.dylib
libomp.dylib
libopenblasp-r0.3.10.dylib
libopenblas.0.dylib
libuuid.so.1.3.0
libcrypto.so.1.0.0
libgcc_s.so.1
libssl.so.1.0.0
libgfortran-ed201abd.so.3.0.0
libopenblasp-r0-382c8f3a.3.5.dev.so
libcrypto.so.1.1
libssl.so.1.1
libffi.so.6.0.4
libgfortran.so.4.0.0
libopenblasp-r0.3.3.so
libquadmath.so.0.0.0
libopenblasp-r0.3.10.so
libgfortran-2e0d59d6.so.5.0.0
libopenblasp-r0-ae94cfde.3.9.dev.so
libquadmath-2d0c479f.so.0.0.0
libz-eb09ad1d.so.1.2.3
libopenblas.IPBC74C7KURV7CB2PKT5Z5FNR3SIBV4J.gfortran-win_amd64.dll
apphelp.dll
crypt32.dll
msasn1.dll
ole32.dll
oleaut32.dll
libifcoremd.dll
libiomp5md.dll
libmmd.dll
mkl_avx512.dll
mkl_core.dll
mkl_intel_thread.dll
mkl_rt.dll
mkl_vml_avx512.dll
tbbmalloc.dll
imagehlp.dll
libcrypto-1_1-x64.dll
libssl-1_1-x64.dll
libblas.dll
libcblas.dll
liblapack.dll
libopenblas.NOIJJG62EMASZI6NYURL6JBKM4EVBGM7.gfortran-win_amd64.dll
libcrypto-1_1.dll
libssl-1_1.dll
libffi-7.dll
mkl_avx2.dll
mkl_vml_avx2.dll
mswsock.dll
```

</details>

Dodgy extensions are:
-  `urandom` (from macOS) which has no suffix.
- `libuuid.so.1.3.0` and quite a few others (from ubuntu) which have version numbers appended to the suffix.

I guess we could fight our way through with regex but it will be ugly and likely find itself too strict to handle future changes or environments I haven't explicitly tested. But I think future compatibility is wistful fantasy anyway thanks to the magic hardcoded list of DLL names.

Reply:
I we were to do a blacklist I'd be worried that there are other cases like `.pbd` that, despite using a wide test matrix, I've still not encountered.

Reply:
Ughhhhh :vomiting_face:

Reply:
Well I think I'll make `.pbd` an exclusion but I imagine we'll be playing whack-a-mole for a long time.

Reply:
Another head-ache to look forward to: The whole macOS Big-Sur crisis may demand that we have to use `ctypes.CDLL(binary)` on all our binaries just to locate them. I guess that would also lead to this problem if we're not careful.

Reply:
> Another head-ache to look forward to: The whole macOS Big-Sur crisis may demand that we have to use `ctypes.CDLL(binary)` on all our binaries just to locate them. I guess that would also lead to this problem if we're not careful.

Doesn't Big Sur hide only system libraries? And we avoid bundling those with built program anyway?

Reply:
As for my test conda environment:

Windows 10 Pro 64-bit, Version 2004 (OS Build 19041.450)
Installed: Anaconda3-2020.07-Windows-x86_64.exe from https://www.anaconda.com

Open Conda prompt:
```
(base) C:\Users\Rok>conda --version
conda 4.8.3

(base) C:\Users\Rok>conda create --name pyinst-numpy python==3.7
...
(base) C:\Users\Rok>conda activate pyinst-numpy
(pyinst-numpy) C:\Users\Rok>conda install numpy
...

(pyinst-numpy) C:\Users\Rok>pip install https://github.com/bwoodsend/pyinstaller/archive/fix-numpy-hooks.zip

(pyinst-numpy) pyinstaller solve.py
...
```

In python 3.7 environment, I now get only error dialog for `libiomp5_md.pdb`, while repeating the above for python 3.8 environment, I get the error dialog for `libcrypto-1_1-x64.pdb`, `libssl-1_1-x64.pdb`, and `libiomp5md.pdb`. (In python 3.7, the other two `.pdb`s are not present).

Reply:
@rokm Do you get these dialogs even after 64e1a71?

Reply:
> Doesn't Big Sur hide only system libraries? And we avoid bundling those with built program anyway?

I really hope so.

Reply:
> @rokm Do you get these dialogs even after [64e1a71](https://github.com/pyinstaller/pyinstaller/commit/64e1a717fb5ef07302da9d94d87c27ad2a810b41)?

Looks OK to me now.